File: jvm/core/src/test/java/org/apache/tvm/contrib/GraphExecutorTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.tvm.contrib;

import org.apache.tvm.Module;
import org.apache.tvm.NDArray;
import org.apache.tvm.Device;
import org.apache.tvm.TestUtils;
import org.apache.tvm.rpc.Client;
import org.apache.tvm.rpc.RPCSession;
import org.apache.tvm.rpc.Server;
import org.junit.BeforeClass;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.Scanner;

import static org.junit.Assert.assertArrayEquals;

public class GraphExecutorTest {
  private final Logger logger = LoggerFactory.getLogger(GraphExecutor.class);
  private static String loadingDir;

  @BeforeClass
  public static void beforeClass() {
    loadingDir = System.getProperty("test.tempdir");
  }

  @Test
  public void test_add_one_local() throws IOException {
    Module libmod = Module.load(loadingDir + File.separator + "graph_addone_lib.so");
    String graphJson = new Scanner(new File(
        loadingDir + File.separator + "graph_addone.json"))
        .useDelimiter("\\Z").next();

    Device dev = Device.cpu();
    GraphModule graph = GraphExecutor.create(graphJson, libmod, dev);

    long[] shape = new long[]{4};
    NDArray arr = NDArray.empty(shape, dev);
    arr.copyFrom(new float[]{1f, 2f, 3f, 4f});

    NDArray out = NDArray.empty(shape, dev);

    graph.setInput("x", arr).run();
    graph.getOutput(0, out);

    assertArrayEquals(new float[]{2f, 3f, 4f, 5f}, out.asFloatArray(), 1e-3f);

    arr.release();
    out.release();
    graph.release();
  }

  @Test
  public void test_add_one_remote() throws IOException {
    if (!Module.enabled("rpc")) {
      logger.warn("RPC is not enabled. Skip.");
      return;
    }

    String libPath = loadingDir + File.separator + "graph_addone_lib.so";
    String graphJson = new Scanner(new File(
        loadingDir + File.separator + "graph_addone.json"))
        .useDelimiter("\\Z").next();

    TestUtils.RefInt port = new TestUtils.RefInt();
    Server server = null;
    try {
      server = TestUtils.startServer(port);
      RPCSession remote = Client.connect("127.0.0.1", port.value);
      Device dev = remote.cpu();

      remote.upload(new File(libPath));
      Module mlib = remote.loadModule("graph_addone_lib.so");

      GraphModule graph = GraphExecutor.create(graphJson, mlib, dev);

      long[] shape = new long[]{4};
      NDArray arr = NDArray.empty(shape, dev);
      arr.copyFrom(new float[]{1f, 2f, 3f, 4f});

      NDArray out = NDArray.empty(shape, dev);

      graph.setInput("x", arr).run();
      graph.getOutput(0, out);

      assertArrayEquals(new float[]{2f, 3f, 4f, 5f}, out.asFloatArray(), 1e-3f);

      arr.release();
      out.release();
      graph.release();
    } finally {
      if (server != null) {
        server.terminate();
      }
    }
  }
}

File: tests/crt/graph_executor_test.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "../../src/runtime/crt/include/tvm/runtime/crt/internal/graph_executor/graph_executor.h"

#include <gtest/gtest.h>

#include "../../src/runtime/crt/include/tvm/runtime/crt/internal/graph_executor/load_json.h"

namespace {

constexpr const char* kJson = R"(
{
  "nodes": [
    {
      "op": "null",
      "name": "x",
      "inputs": []
    },
    {
      "op": "null",
      "name": "p0",
      "inputs": []
    },
    {
      "op": "tvm_op",
      "name": "tvmgen_default_fused_add",
      "attrs": {
        "num_outputs": "1",
        "num_inputs": "2",
        "flatten_data": "0",
        "func_name": "tvmgen_default_fused_add",
        "hash": "a2b7e0a88031366c"
      },
      "inputs": [
        [
          0,
          0,
          0
        ],
        [
          1,
          0,
          0
        ]
      ]
    }
  ],
  "arg_nodes": [0, 1],
  "heads": [
    [
      2,
      0,
      0
    ]
  ],
  "attrs": {
    "dltype": [
      "list_str",
      [
        "float32",
        "float32",
        "float32"
      ]
    ],
    "device_index": [
      "list_int",
      [1, 1, 1]
    ],
    "storage_id": [
      "list_int",
      [0, 1, 2]
    ],
    "shape": [
      "list_shape",
      [
        [10, 5],
        [1, 5],
        [10, 5]
      ]
    ]
  },
  "node_row_ptr": [0, 1, 2, 3]
}
)";

// Check a JSON graph can be loaded.
TEST(TVMGraphExecutor_Load, Parse) {
  JSONReader reader;
  tvm_crt_error_t err = JSONReader_Create(kJson, &reader);
  EXPECT_EQ(err, kTvmErrorNoError);
  TVMGraphExecutor executor;
  memset(&executor, 0, sizeof(executor));
  int status = TVMGraphExecutor_Load(&executor, &reader);
  EXPECT_EQ(status, 0);
  EXPECT_EQ(executor.nodes_count, 3);
}

}  // namespace
